controllers/helmrepository_controller_oci.go

```diff
@@ -0,0 +1,362 @@
+	return sreconcile.ResultEmpty, nil
+}
+
+// notify emits notification related to the reconciliation.
+func (r *HelmRepositoryOCIReconciler) notify(oldObj, newObj *sourcev1.HelmRepository, res sreconcile.Result, resErr error) {
+	// Notify successful recovery from any failure.
+	if resErr == nil && res == sreconcile.ResultSuccess {
+		if sreconcile.FailureRecovery(oldObj, newObj, helmRepositoryOCIFailConditions) {
+			r.Eventf(newObj, corev1.EventTypeNormal,
+				meta.SucceededReason, "Helm repository %q has been successfully reconciled", newObj.Name)
+		}
+	}
+}
+
+func (r *HelmRepositoryOCIReconciler) reconcile(ctx context.Context, obj *sourcev1.HelmRepository, reconcilers []helmRepositoryOCIReconcileFunc) (sreconcile.Result, error) {
+	oldObj := obj.DeepCopy()
+
+	// Mark as reconciling if generation differs.
+	if obj.Generation != obj.Status.ObservedGeneration {
+		conditions.MarkReconciling(obj, "NewGeneration", "reconciling new object generation (%d)", obj.Generation)
+	}
+
+	// Run the sub-reconcilers and build the result of reconciliation.
+	var res sreconcile.Result
+	var resErr error
+	for _, rec := range reconcilers {
+		recResult, err := rec(ctx, obj)
+		// Exit immediately on ResultRequeue.
+		if recResult == sreconcile.ResultRequeue {
+			return sreconcile.ResultRequeue, nil
+		}
+		// If an error is received, prioritize the returned results because an
+		// error also means immediate requeue.
+		if err != nil {
+			resErr = err
+			res = recResult
+			break
+		}
+		// Prioritize requeue request in the result for successful results.
+		res = sreconcile.LowestRequeuingResult(res, recResult)
+	}
+
+	r.notify(oldObj, obj, res, resErr)
+
+	return res, resErr
+}
+
+func (r *HelmRepositoryOCIReconciler) reconcileSource(ctx context.Context, obj *sourcev1.HelmRepository) (sreconcile.Result, error) {
+	var logOpts []registry.LoginOption
+	// Configure any authentication related options
+	if obj.Spec.SecretRef != nil {
+		// Attempt to retrieve secret
+		name := types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      obj.Spec.SecretRef.Name,
+		}
+		var secret corev1.Secret
+		if err := r.Client.Get(ctx, name, &secret); err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to get secret '%s': %w", name.String(), err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		// Construct actual options
+		logOpt, err := loginOptionFromSecret(secret)
+		if err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to configure Helm client with secret data: %w", err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			// Return err as the content of the secret may change.
+			return sreconcile.ResultEmpty, e
+		}
+
+		logOpts = append(logOpts, logOpt)
```

Comment:
The `logOpt` needs to be checked for nil before appending here. If the username and password are empty in the secret, `nil, nil` are returned from the `loginOptionFromSecret()`. When a slice with <nil> is used later for logging in, it results in panic.

Reply:
Also, maybe the RFC needs to be updated. I was following the example from RFC for the secret which expects it to be a docker-registry secret, and discovered this issue.
Refer: https://github.com/fluxcd/flux2/tree/e3e01cb5da3ae248b4e2a1e5a600cfbcb0ad82a3/rfcs/0002-helm-oci#story-1

Reply:
I'm working on this right now.

Reply:
Draft PR: https://github.com/fluxcd/source-controller/pull/725